kore/src/Kore/Step/Simplification/Condition.hs

```diff
@@ -120,49 +123,71 @@ simplify SubstitutionSimplifier{simplifySubstitution} sideCondition =
+others are true.
+This procedure is applied until the conjunction stabilizes.
 -}
 simplifyPredicates ::
-    forall simplifier.
-    HasCallStack =>
     MonadSimplify simplifier =>
     SideCondition RewritingVariableName ->
-    Predicate RewritingVariableName ->
+    MultiAnd (Predicate RewritingVariableName) ->
     LogicT simplifier (Condition RewritingVariableName)
-simplifyPredicates
-    sideCondition
-    (toList . MultiAnd.fromPredicate -> predicates) =
-        State.execStateT (worker predicates) Condition.top
-            >>= markConjunctionSimplified
-      where
-        worker ::
-            [Predicate RewritingVariableName] ->
-            StateT
-                (Condition RewritingVariableName)
-                (LogicT simplifier)
-                ()
-        worker [] = return ()
-        worker (pred' : rest) = do
-            condition <- State.get
-            let otherConds =
-                    SideCondition.andCondition
-                        sideCondition
-                        (mkOtherConditions condition rest)
-            result <-
-                simplifyPredicate otherConds pred'
-                    & lift
-            State.put (Condition.andCondition condition result)
-            worker rest
-
-        mkOtherConditions (Condition.toPredicate -> alreadySimplified) rest =
-            from @_ @(Condition _)
-                . MultiAnd.toPredicate
-                . MultiAnd.make
-                $ alreadySimplified : rest
+simplifyPredicates sideCondition original = do
+    let predicates =
+            SideCondition.simplifyConjunctionByAssumption original
+                & fst . extract
+    simplified <-
+        simplifyPredicatesWithAssumptions
+            sideCondition
+            (toList predicates)
+    let simplifiedPredicates =
+            from @(Condition _) @(MultiAnd (Predicate _)) simplified
+    if original == simplifiedPredicates
+        then return (Condition.markSimplified simplified)
+        else simplifyPredicates sideCondition simplifiedPredicates
 
-        markConjunctionSimplified =
-            return . Lens.over (field @"predicate") Predicate.markSimplified
+{- | Simplify a conjunction of predicates by simplifying each one
+under the assumption that the others are true.
+-}
+simplifyPredicatesWithAssumptions ::
+    forall simplifier.
+    MonadSimplify simplifier =>
+    SideCondition RewritingVariableName ->
+    [Predicate RewritingVariableName] ->
+    LogicT
+        simplifier
+        (Condition RewritingVariableName)
+simplifyPredicatesWithAssumptions _ [] = return Condition.top
+simplifyPredicatesWithAssumptions sideCondition predicates@(_ : rest) = do
+    let predicatesWithUnsimplified =
+            zip predicates $
+                scanr ((<>) . MultiAnd.singleton) MultiAnd.top rest
+    State.execStateT
+        ( traverse
+            simplifyWithAssumptions
+            predicatesWithUnsimplified
+        )
+        Condition.top
```

Comment:
I am still confused why this is `execStateT` instead of `evalStateT`. Can you remind my why the return value is the "state" instead of the `SideCondition`?

Reply:
After giving this some thought, I think that in the end the result will be similar: with `evalStateT` we'd end up with a list of simplified predicates which we can fold into a `Condition` which will be the final result, right? But I don't think we need to do that, since we already accumulate the result in the state (which we also need as a `SideCondition` during simplification). Does this make sense?

Reply:
Yes, I think that's fine. In that case, we should use `traverse_` here instead of `traverse`.